foundry.md

```diff
@@ -295,9 +295,28 @@ However, there is no change on the state and therefore this rule just skips the
          <localMem> LM </localMem>
       requires CHEAT_ADDR ==Int #address(FoundryCheat)
        andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 3327641368 // selector ( "label(address,string)" )
+    [priority(40)]
 ```
 
+#### `addr` - Computes the address for a given private key.
+
+```
+function addr(uint256 privateKey) external returns (address);
+```
+
+`call.addr` will match when the `addr` function is called at the [Foundry cheatcode address](https://book.getfoundry.sh/cheatcodes/#cheatcodes-reference).
+This rule takes `uint256` value using `#asWord(#range(LM, ARGSTART +Int 4, 32)` and computes the address for a given private key updating the `<output>` cell.
+
 ```k
-endmodule
+    rule [call.addr]:
+          <k> CALL _ CHEAT_ADDR 0 ARGSTART _ARGWIDTH RETSTART RETWIDTH => 1 ~> #push ~> #setLocalMem RETSTART RETWIDTH #bufStrict(32, #addrFromPrivateKey(#unparseByteStack(#range(LM, ARGSTART +Int 4, 32)))) ... </k>
```

Comment:
Here it is required that no ether is sent. On the other hand, Foundry *does* allow you to send ether when calling `vm.addr()`, which admittedly does not make much sense, but if we want to be compatible with Foundry we should probably behave the same.

Reply:
How is it possible to send ether when calling `vm.addr()`? I just changed the rule, but would like to make a test for that.

Reply:
Usually you can just do `foo.bar{value: 1 ether}(...)` to send ether with a call. However, for `vm.addr()` this does not work directly because it is not `payable`. Therefore one needs to create another contract/interface that changes the signature of `addr` accordingly:
```solidity
interface PayableVM {
    function addr(uint256 p) external payable returns(address);
}
```
In a test function you can then cast `vm` to `PayableVM` and send ether like so: `vm.addr{value: 1 ether}(...)`

Reply:
@david-k I used `#checkCall` and `#transferFunds` to support the ether transfers. Please take a look and let me know your thoughts.

Reply:
So I did some more tests and it turns out that while you can send ether the `vm`, it seems to get lost on the way, i.e., it does not influence `vm.balance`:
```solidity
PayableVM(address(vm)).addr{value: 4}(23);
console.log(address(vm).balance); // Outputs "0" and not "4"
```
According to this it seems we don't actually need `#transferFunds`. Thus, we can just use a wildcard `_`  for `VALUE` and ignore the ether that is sent. I'll ask on Discord how exactly they process calls to the VM, maybe there are other edge-cases. Sorry for all the back-and-forth!

Reply:
Ahh ok! It was useful for me to learn anyway, so thanks!